cmake/Modules/Packages/MSCG.cmake

```diff
@@ -12,34 +12,48 @@ if(DOWNLOAD_MSCG)
   mark_as_advanced(MSCG_URL)
   mark_as_advanced(MSCG_MD5)
 
-  include(ExternalProject)
-  ExternalProject_Add(mscg_build
-    URL     ${MSCG_URL}
-    URL_MD5 ${MSCG_MD5}
-    SOURCE_SUBDIR src/CMake
-    CMAKE_ARGS ${CMAKE_REQUEST_PIC} ${EXTRA_MSCG_OPTS}
-               -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
-               -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
-               -DCMAKE_Fortran_COMPILER=${CMAKE_Fortran_COMPILER}
-               -DBLAS_LIBRARIES=${BLAS_LIBRARIES} -DLAPACK_LIBRARIES=${LAPACK_LIBRARIES}
-               -DCMAKE_INSTALL_PREFIX=<INSTALL_DIR>
-               -DCMAKE_BUILD_TYPE=${CMAKE_BUILD_TYPE}
-               -DCMAKE_MAKE_PROGRAM=${CMAKE_MAKE_PROGRAM}
-               -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TOOLCHAIN_FILE}
-    BUILD_COMMAND ${CMAKE_COMMAND} --build . --target mscg
-    INSTALL_COMMAND ""
-    BUILD_BYPRODUCTS <BINARY_DIR>/libmscg.a
-    )
-  ExternalProject_get_property(mscg_build BINARY_DIR)
-  ExternalProject_get_property(mscg_build SOURCE_DIR)
-  file(MAKE_DIRECTORY ${SOURCE_DIR}/src)
-  add_library(LAMMPS::MSCG UNKNOWN IMPORTED)
-  set_target_properties(LAMMPS::MSCG PROPERTIES
-    IMPORTED_LOCATION "${BINARY_DIR}/libmscg.a"
-    INTERFACE_INCLUDE_DIRECTORIES "${SOURCE_DIR}/src"
-    INTERFACE_LINK_LIBRARIES "${LAPACK_LIBRARIES}")
-  target_link_libraries(lammps PRIVATE LAMMPS::MSCG)
-  add_dependencies(LAMMPS::MSCG mscg_build)
+  # always compile a static lib but with position independent code
+  # make a copy of current settings for later use
+  set(OLD_SHARED_LIBS ${BUILD_SHARED_LIBS})
```

Comment:
I think the more elegant solution would be to define a CMake function and overwrite these 4 variables in there. Variable values are scoped inside the function once the function ends variables are restored.

Also the download & unpack part is used 3 times already, so it might be worth a function anyhow.

Reply:
Thanks for the suggestion. Was trying to work around some limitations for "ExternalProject_add()":
- it doesn't seem to propagate the current configuration with multi-configuration build systems (e.g. ninja multi-config)
- it doesn't seem to be possible to pass BLAS or LAPACK library definitions unless it is a single library file. With "flexiblas" on Fedora, building any such external project breaks with an error like "unknown command '-ldl'". So it seems in the long run, the only clean way to have external CMake based libraries included is by using "add_subdirectory()", but then the issue is that we want the library code to be included in liblammps.so.0, so that there is less of a mess of n+1 shared libraries that need to be kept alongside liblammps.so.0. hence the hackery with turning shared/static on/off. not a very satisfactory solution...

Not sure if this is the right way to move forward at all. It just came out of some experiments I did on a different branch and I didn't want to "lose" the changes for multi-config and passing flexiblas style LAPACK libs.